### ISSUE TITLE
Arguments of FindRoot are not evaluated

### ISSUE THREAD

**Issue by soehms** [mathics/Mathics#1235]
Using predefined arguments gives:

```
In[4]:= v1 := Exp[x] - 3x == 0
In[5]:= v2 := {x, 2}
In[6]:= FindRoot[v1, v2]
Out[6]= FindRoot[v1, v2]
```
Using explicit arguments works:

```
In[7]:= FindRoot[Exp[x] - 3x == 0, {x, 2}]
Out[7]= {x -> 1.51213}
```

Shouldn't `Out[6]` be the same as `Out[7]`?

**Comment by rocky** [comment]
#1236 starts to address some of this, but unless someone else gets interested in it (@soehms maybe you can start here and work more on it?), from my standpoint this will have to wait until what is needed for FeynCalc to work is more complete, since that probably will have more of a public impact. 

But here is what is going on. 

Currently the function form that is matched is `FindRoot[f_, {x_, x0_}]`  which is expecting the second argument to be written out as a list of two items.  So it is not so much an _evaluation_ problem so much as a pattern matching problem which is causing the code to handle `FindRoot[]` not to get triggered. Or rather that is just _one_ of the problems. 

In the FindRoot-expansion` branch I created to start addressing this, I added code to match the situation here where the second parameter is a variable containing a list. 

However there is transformation rule that looks like this: `FindRoot[lhs_ == rhs_, {x_, xs_}]": "FindRoot[lhs - rhs, {x, xs}]`.

In other words, a function whose top-level operation is `Equal[]` in them are changed into  subtraction.

The code I wrote detects that but isn't doing the transformation in code properly. It might be a mistake I've made. But as I said, that someone else should look at and move forward at some point. 

So to sum up... 

```
n[4]:= v1 := Exp[x] - 3x  (* no: "== 0" which is the same thing *)
In[5]:= v2 := {x, 2}
In[6]:= FindRoot[v1, v2]
```
now works. Although adding the `== 0` doesn't yet.  A different error, a recursion error, is given for this.

**Comment by rocky** [comment]
I should also mention that mpmath also provides a `findroot()` function and it might be that the iterative code we have here.
Also [sympy's nsolve seems to use the mpmath findroot](https://docs.sympy.org/latest/modules/solvers/solvers.html). We may hook into that via the Mathics `Solve` function.

**Comment by mmatera** [comment]
> #1236 starts to address some of this, but unless someone else gets interested in it (@soehms maybe you can start here and work more on it?), from my standpoint this will wait until what is needed for FeynCalc to work is more complete.
> 
> But here is what is going on.
> 
> Currently the function form that is matched is `FindRoot[f_, {x_, x0_}]` which is expecting the second argument to be written out as a list of two items. So it is not so much an _evaluation_ problem so much as a pattern matching problem which is causing the code to handle `FindRoot[]` not to get triggered. Or rather that is just _one_ of the problems.
> 
> In the `FindRoot-expansion` branch I created to start addressing this, I added code to match the situation here where the second parameter is a variable containing a list.
> 
> However there is transformation rule that looks like this: `FindRoot[lhs_ == rhs_, {x_, xs_}]": "FindRoot[lhs - rhs, {x, xs}]`.
> 
> In other words, a function whose top-level operation is `Equal[]` is changed into subtraction.
> 
> The code I wrote detects that but isn't doing the transformation in code properly. It might be a mistake I've made. But as I said, that someone else should look at and move forward at some point.
> 
> So to sum up...
> 
> ```
> n[4]:= v1 := Exp[x] - 3x  (* no: "== 0" which is the same thing *)
> In[5]:= v2 := {x, 2}
> In[6]:= FindRoot[v1, v2]
> ```
> 
> now works. Although adding the `== 0` doesn't yet. A different error, a recursion error, is given for this.

I can look at this over the weekend.

**Comment by soehms** [comment]
Thanks a lot for looking at this.

> @soehms maybe you can start here and work more on it?

I fear I have to learn much more about Mathics (and even Mathematica) before I can do that.

> Currently the function form that is matched is FindRoot[f_, {x_, x0_}] which is expecting the second argument to be written out as a list of two items ....

Indeed, I suspected something like this, but astonishingly the following works:

```
In[4]:= u1 := {x^2 - 3y == 3, 2x - y == 1}
In[5]:= u2 := {x, y}
In[6]:= Solve[u1, u2]
Out[6]= {{x -> 0, y -> -1}, {x -> 6, y -> 11}}
```

BTW: These examples are doctests of [Sage's interface to Mathematica](https://doc.sagemath.org/html/en/reference/interfaces/sage/interfaces/mathematica.html?highlight=mathematica#module-sage.interfaces.mathematica). I've tried to transform a clone of [its code](https://github.com/sagemath/sage/blob/develop/src/sage/interfaces/mathematica.py) into an interface to Mathics (using your library functionality directly instead of `Pexpect`).

Apart from minor differences in output formatting the issue explained here was the first failing test (line 175). The next one is `FindMinimum[x^3 - 6x^2 + 11x - 5, {x,3}]` where it seems that this function is completely missing in Mathics. But I really don't know if these doctests are currently checked against recent Mathematica releases.

But the most curious observation I made (see the code line 265) seems to be related to the issue of [this comment](https://github.com/mathics/Mathics/issues/1169#issuecomment-806532433) in #1169. To reproduce it you have to run Mathics with Sage:

```
~/devel/sage$ sage local/lib/python3.9/site-packages/mathics/main.py

Mathics 2.0.1.dev
on CPython 3.9.2 (default, Mar 19 2021, 22:23:28)
using SymPy 1.7, mpmath 1.2.1, numpy 1.19.5, cython 0.29.21

Copyright (C) 2011-2021 The Mathics Team.
This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under certain conditions.
See the documentation for the full license.

Quit by evaluating Quit[] or by pressing CONTROL-D.

In[1]:= N[Pi, 37]
Out[1]= 3.141592653589793238462643383279502884

In[2]:= N[Pi, 38]
OverflowError: Python int too large to convert to C long

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/sebastian/devel/sage/local/lib/python3.9/site-packages/mathics/main.py", line 382, in <module>
    sys.exit(main())

...............

  File "mathics/core/rules.py", line 126, in mathics.core.rules.BuiltinRule.do_replace
  File "/home/sebastian/devel/sage/local/lib/python3.9/site-packages/mathics/builtin/constants.py", line 80, in apply_N
    return self.get_constant(precision, evaluation, preference)
  File "/home/sebastian/devel/sage/local/lib/python3.9/site-packages/mathics/builtin/constants.py", line 136, in get_constant
    return PrecisionReal(sympy.Float(str(value), d))

  File "sage/libs/mpmath/ext_main.pyx", line 1685, in sage.libs.mpmath.ext_main.mpf_base.__str__ (build/cythonized/sage/libs/mpmath/ext_main.c:21574)

  File "/home/sebastian/devel/sage/local/lib/python3.9/site-packages/mpmath/libmp/libmpf.py", line 1242, in to_str
    sign, digits, exponent = to_digits_exp(s, dps+3)
  File "/home/sebastian/devel/sage/local/lib/python3.9/site-packages/mpmath/libmp/libmpf.py", line 1200, in to_digits_exp
    digits = numeral(sd, base=10, size=dps)
  File "/home/sebastian/devel/sage/local/lib/python3.9/site-packages/mpmath/libmp/libintmath.py", line 166, in numeral_python
    A, B = divmod(n, base**half)
SystemError: <built-in function divmod> returned a result with an error set
```

But probably this is a bug in Sage and I will investigate this in more detail. Anyway, since I'm not familiar with Mpmath (and the Sage interface to it) any hints are welcome.

**Comment by mmatera** [comment]
This should be addressed by PR #1277

**Comment by soehms** [comment]
> But the most curious observation I made ......

This is [Trac ticket #29009](https://trac.sagemath.org/29009).

**Comment by rocky** [comment]
The way this kind of thing should be (is?) handled in Mathics is that when somethng raises an exception, it is caught and another fallback method is used. And separatel a user can specify which method to use, the same as we have for `N[]`, so you can say `N[Pi, Method->"mpmath"] or alternatively N[Pi, Method->"sympy"].

I would hope sympy does this too.

**Comment by soehms** [comment]
>The way this kind of thing should be (is?) handled in Mathics s that when something raises an exception ....

My experience is that you cannot catch this kind of `OverfowError`. 

If there are choices of implementation in Sage the default choice is implemented explicitly in most cases, especially if it is a matter of performance. But I think, fallbacks in cases of exceptions are used, as well. 

Anyway the above exception was caused by a Python3 incompatibility of a line of code written in the spirit of Python2 (the switch to Python3 in Sage is not long ago). It will be fixed in the first beta of version 9.4.

BTW: Concerning `N[]` I'm getting the following differences to the results (marked as `Expected`) that once have been produced using Sage's Mathematica Interface  (see [my comment on 7.th of April](https://github.com/mathics/Mathics/issues/1235#issuecomment-814899766)):


```
File "src/sage/interfaces/mathics.py", line 350, in sage.interfaces.mathics
Failed example:
    mathics('Pi/2').N(10)           # optional -- mathics
Expected:
    1.5707963268
Got:
    1.570796327
**********************************************************************
File "src/sage/interfaces/mathics.py", line 352, in sage.interfaces.mathics
Failed example:
    mathics('Pi').N(10)             # optional -- mathics
Expected:
    3.1415926536
Got:
    3.141592654
**********************************************************************
File "src/sage/interfaces/mathics.py", line 354, in sage.interfaces.mathics
Failed example:
    mathics('Pi').N(50)             # optional -- mathics
Expected:
    3.14159265358979323846264338327950288419716939937511
Got:
    3.1415926535897932384626433832795028841971693993751
```

What is the reason for this?

**Comment by mmatera** [comment]
In MMA, the digits behind the decimal point are also counted as precision digits:

``
In[1] = N[10*Pi,10]

Out[1] = 31.41592654
``

**Comment by soehms** [comment]
> BTW: These examples are doctests of Sage's interface to Mathematica. I've tried to transform a clone of its code into an interface to Mathics (using your library functionality directly instead of Pexpect).

My first attempt can be seen on [Trac ticket #31778](https://trac.sagemath.org/ticket/31778)!

I've seen that the issue reported here has been fixed in the current development version. Therefore, I close this ticket.
